src/jsCore.ts

```diff
@@ -0,0 +1,498 @@
+      version: details.eventData.version,
+      description: details.eventData.documentDescription,
+      expiry: details.eventData.expiry,
+    });
+    tracker.track(built.event, details.contexts.concat(built.context));
+  });
+  return <Promise<void>>Promise.resolve();
+}
+
+function trackConsentWithdrawnEvent(details: {
+  tracker: string | null;
+  eventData: ConsentWithdrawnProps;
+  contexts: EventContext[];
+}): Promise<void> {
+  forTracker(details.tracker, (tracker) => {
+    const built = buildConsentWithdrawn({
+      all: details.eventData.all,
+      id: details.eventData.documentId,
+      version: details.eventData.version,
+      description: details.eventData.documentDescription,
+    });
+    tracker.track(built.event, details.contexts.concat(built.context));
+  });
+  return <Promise<void>>Promise.resolve();
+}
+
+function trackEcommerceTransactionEvent(details: {
+  tracker: string | null;
+  eventData: EcommerceTransactionProps;
+  contexts: EventContext[];
+}): Promise<void> {
+  forTracker(details.tracker, (tracker) => {
+    tracker.track(
+      buildEcommerceTransaction({
+        orderId: details.eventData.orderId,
+        total: details.eventData.totalValue,
+        affiliation: details.eventData.affiliation,
+        tax: details.eventData.taxValue,
+        shipping: details.eventData.shipping,
+        city: details.eventData.city,
+        state: details.eventData.state,
+        country: details.eventData.country,
+        currency: details.eventData.currency,
+      }),
+      details.contexts
+    );
+  });
+  return <Promise<void>>Promise.resolve();
+}
+
+function trackDeepLinkReceivedEvent(details: {
+  tracker: string | null;
+  eventData: DeepLinkReceivedProps;
+  contexts: EventContext[];
+}): Promise<void> {
+  return trackSelfDescribingEvent({
+    tracker: details.tracker,
+    eventData: {
+      schema:
+        'iglu:com.snowplowanalytics.mobile/deep_link_received/jsonschema/1-0-0',
+      data: details.eventData,
+    },
+    contexts: details.contexts,
+  });
+}
+
+function trackMessageNotificationEvent(details: {
+  tracker: string | null;
+  eventData: MessageNotificationProps;
+  contexts: EventContext[];
+}): Promise<void> {
+  return trackSelfDescribingEvent({
+    tracker: details.tracker,
+    eventData: {
+      schema:
+        'iglu:com.snowplowanalytics.mobile/message_notification/jsonschema/1-0-0',
+      data: details.eventData,
+    },
+    contexts: details.contexts,
+  });
+}
+
+function removeGlobalContexts(details: { tracker: string; removeTag: string }): Promise<void> {
+  details;
+  return <Promise<void>>Promise.reject('Not implemented');
+}
+
+function addGlobalContexts(details: {
+  tracker: string;
+  addGlobalContext: GlobalContext;
+}): Promise<void> {
+  details;
+  return <Promise<void>>Promise.reject('Not implemented');
+}
+
+function setUserId(details: { tracker: string; userId: string | null }): Promise<void> {
+  if (details.userId != null) {
```

Comment:
I think that is okay (unless I misunderstand something) because the `subject.ts` file is called before the `JSSnowplowTracker` is called – it calls the native functions. So the checks for `null` that are there also apply for the `JSSnowplowTracker`.

Reply:
Ohh, I see, that's true, setting `null` in the native tracker resets the property but here it is ignored. Would be better if it behaved the same way, but unfortunately the JS tracker core doesn't accept `null` and doesn't provide a way to reset the values. We could do something custom in this jsCore wrapper but I would prefer keep it very simple at this time.

Reply:
AFAIR if you want to unset userid in JS tracker you pass an empty string. Bit odd/jank, probably should change, but that's the behaviour (at least last I checked, which is a while ago) fwiw

Reply:
Good shout, Colm! I think you are right, at least [this check here](https://github.com/snowplow/snowplow-javascript-tracker/blob/50fb7ea55b1bc7d41376537bfb78d512fc2e7683/libraries/tracker-core/src/payload.ts#L121) suggests that empty strings are ignored when adding to payload. Will change it to set empty strings, thanks!